arangod/Cluster/ServerState.cpp

```diff
@@ -647,23 +669,30 @@ bool ServerState::registerAtAgencyPhase2(AgencyComm& comm) {
   TRI_ASSERT(!_id.empty() && !_myEndpoint.empty());
 
   while (!application_features::ApplicationServer::isStopping()) {
-    VPackBuilder builder;
-    try {
-      VPackObjectBuilder b(&builder);
-      builder.add("endpoint", VPackValue(_myEndpoint));
-      builder.add("advertisedEndpoint", VPackValue(_advertisedEndpoint));
-      builder.add("host", VPackValue(getHost()));
-      builder.add("version", VPackValue(rest::Version::getNumericServerVersion()));
-      builder.add("versionString", VPackValue(rest::Version::getServerVersion()));
-      builder.add("engine", VPackValue(EngineSelectorFeature::engineName()));
-    } catch (...) {
-      LOG_TOPIC("de625", FATAL, arangodb::Logger::CLUSTER) << "out of memory";
-      FATAL_ERROR_EXIT();
-    }
-
-    auto result = comm.setValue(currentServersRegisteredPref + _id, builder.slice(), 0.0);
+    std::vector<std::shared_ptr<VPackBuilder>> builders;
+
+    auto buildSetOp = [this, &builders](char const* url, VPackValue const& v) {
+      builders.emplace_back(new VPackBuilder);
+      builders.back()->add(v);
+      return AgencyOperation(currentServersRegisteredPref + _id + "/" + url,
+                             AgencyValueOperationType::SET, builders.back()->slice());
+    };
+
+    AgencyWriteTransaction trx(
+        {buildSetOp("endpoint", VPackValue(_myEndpoint)),
+         buildSetOp("advertisedEndpoint", VPackValue(_advertisedEndpoint)),
+         buildSetOp("host", VPackValue(getHost())),
+         buildSetOp("version", VPackValue(rest::Version::getNumericServerVersion())),
+         buildSetOp("versionString", VPackValue(rest::Version::getServerVersion())),
+         buildSetOp("engine", VPackValue(EngineSelectorFeature::engineName())),
+         AgencyOperation(currentServersRegisteredPref + _id + "/rebootId",
+                         AgencySimpleOperationType::INCREMENT_OP),
+         AgencyOperation("Current/Version", AgencySimpleOperationType::INCREMENT_OP)});
+
```

Comment:
BIG NOTE: I put this in here out of my own intuition that the version should be incremented when I change the document. This *was not previously there* so it would be helpful if someone more familiar with cluster stuff could check that this is OK (i.e. @maierlars @kvahed or @neunhoef)

Reply:
Actually I think it is a bug, that it was NOT increased before.
The contract is:
Modify Current => CurrentVersion++
This was violated here, this could have the side effect that other servers for a short period did not see this server (they do not bother to reload current)
I guess this will be solved by the server itself, as in the first maintenance run it will figure out that it needs to create databases (at least _system) which will modify current.

Reply:
True